coqlspclient/coq_lsp_client.py

```diff
@@ -0,0 +1,197 @@
+        if uri not in self.file_progress:
+            self.file_progress[uri] = [coqFileProgressKind]
+        else:
+            self.file_progress[uri].append(coqFileProgressKind)
+
+    def __wait_for_operation(self):
+        timeout = self.lsp_endpoint.timeout
+        while not self.__completed_operation and timeout > 0:
+            if self.lsp_endpoint.shutdown_flag:
+                raise ResponseError(ErrorCodes.ServerQuit, "Server quit")
+            time.sleep(0.1)
+            timeout -= 0.1
+
+        if timeout <= 0:
+            self.shutdown()
+            self.exit()
+            raise ResponseError(ErrorCodes.ServerQuit, "Server quit")
+
+    def didOpen(self, textDocument: TextDocumentItem):
+        """Open a text document in the server.
+
+        Args:
+            textDocument (TextDocumentItem): Text document to open
+        """
+        self.__completed_operation = False
+        super().didOpen(textDocument)
+        self.__wait_for_operation()
+
+    def didChange(
+        self,
+        textDocument: VersionedTextDocumentIdentifier,
+        contentChanges: list[TextDocumentContentChangeEvent],
+    ):
+        """Submit changes on a text document already open on the server.
+
+        Args:
+            textDocument (VersionedTextDocumentIdentifier): Text document changed.
+            contentChanges (list[TextDocumentContentChangeEvent]): Changes made.
+        """
+        self.__completed_operation = False
+        super().didChange(textDocument, contentChanges)
+        self.__wait_for_operation()
+
+    def proof_goals(
+        self, textDocument: TextDocumentIdentifier, position: Position
+    ) -> Optional[GoalAnswer]:
+        """Get proof goals and relevant information at a position.
+
+        Args:
+            textDocument (TextDocumentIdentifier): Text document to consider.
+            position (Position): Position used to get the proof goals.
+
+        Returns:
+            GoalAnswer: Contains the goals at a position, messages associated
+                to the position and if errors exist, the top error at the position.
+        """
+        result_dict = self.lsp_endpoint.call_method(
+            "proof/goals", textDocument=textDocument, position=position
+        )
+        result_dict["textDocument"] = VersionedTextDocumentIdentifier(
+            **result_dict["textDocument"]
+        )
+        result_dict["position"] = Position(
+            result_dict["position"]["line"], result_dict["position"]["character"]
+        )
+
+        if result_dict["goals"] is not None:
+            result_dict["goals"] = GoalConfig.parse(result_dict["goals"])
+
+        for i, message in enumerate(result_dict["messages"]):
+            if not isinstance(message, str):
+                if message["range"]:
+                    message["range"] = Range(**message["range"])
+                result_dict["messages"][i] = Message(**message)
+
+        return GoalAnswer(**result_dict)
```

Comment:
I think it makes sense to leave the handling of result_dict inside the GoalAnswer constructor. This is done below with the FlecheDocument constructor in get_document.